x-pack/legacy/plugins/actions/server/lib/license_state.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { i18n } from '@kbn/i18n';
+import { Observable, Subscription } from 'rxjs';
+import { ILicense, LICENSE_CHECK_STATE } from '../../../../../plugins/licensing/common/types';
+import { assertNever } from '../../../../../../src/core/utils';
+import { PLUGIN } from '../constants/plugin';
+
+export interface ActionsLicenseInformation {
+  showAppLink: boolean;
+  enableAppLink: boolean;
+  message: string;
+}
+
+export class LicenseState {
+  private licenseInformation: ActionsLicenseInformation = this.checkLicense(undefined);
+  private subscription: Subscription | null = null;
+
+  private updateInformation(license: ILicense | undefined) {
+    this.licenseInformation = this.checkLicense(license);
+  }
+
+  public start(license$: Observable<ILicense>) {
+    this.subscription = license$.subscribe(this.updateInformation.bind(this));
+  }
+
+  public stop() {
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+
+  public getLicenseInformation() {
+    return this.licenseInformation;
+  }
+
+  public checkLicense(license: ILicense | undefined): ActionsLicenseInformation {
+    if (!license || !license.isAvailable) {
+      return {
+        showAppLink: true,
+        enableAppLink: false,
+        message: i18n.translate(
+          'xpack.actions.serverSideErrors.unavailableLicenseInformationErrorMessage',
+          {
+            defaultMessage:
+              'Actions is unavailable - license information is not available at this time.',
+          }
+        ),
+      };
+    }
+
+    // TODO: feature validation?
+    // const actionsFeature = license.getFeature('actions');
+
+    const check = license.check(PLUGIN.ID, PLUGIN.MINIMUM_LICENSE_REQUIRED);
+
+    switch (check.state) {
+      case LICENSE_CHECK_STATE.Expired:
+        return {
+          showAppLink: true,
+          enableAppLink: false,
+          message: check.message || '',
+        };
+      case LICENSE_CHECK_STATE.Invalid:
+      case LICENSE_CHECK_STATE.Unavailable:
+        return {
+          showAppLink: false,
+          enableAppLink: false,
+          message: check.message || '',
+        };
+      case LICENSE_CHECK_STATE.Valid:
+        return {
+          showAppLink: true,
+          enableAppLink: true,
+          message: '',
+        };
+      default:
+        return assertNever(check.state);
+    }
+  }
+}
+
+export function verifyApiAccessFactory(licenseState: LicenseState) {
```

Comment:
I think this would be cleaner if we don't return an inner function and rather use `bind` to pre-bake the `licenseState` where we want to.

For example:
```
function verifyApiAccessFactory(licenseState) {
  throw new Error(licenseState.state);
}

const state = { state: 'invalid' };

// this gives us a "prebaked" version, so whjen we call the function, we don't need to provide the state 
const prebakedWithLicenseState = verifyApiAccessFactory.bind(state);

console.log(`this will log, as we haven't executed yet`);

// we can now call `prebakedWithLicenseState` when we need to.
prebakedWithLicenseState();

console.log(`this won't log as the error throws in the line above`);
```